tests/api/test_share.py

```diff
@@ -17,139 +15,226 @@ def cpltz(l):
+        status=dataall.api.constants.ShareObjectStatus.Draft.value
+    )
+
+
+@pytest.fixture(scope='module')
+def share_item2_pa(
+        share_item: typing.Callable,
+        share2: dataall.db.models.ShareObject,
+        table2: dataall.db.models.DatasetTable
+) -> dataall.db.models.ShareObjectItem:
+    yield share_item(
+        share=share2,
+        table=table2,
+        status=dataall.api.constants.ShareItemStatus.PendingApproval.value
+    )
+
+
+@pytest.fixture(scope='module')
+def share_item2_shared(
+        share_item: typing.Callable,
+        share2: dataall.db.models.ShareObject,
+        table2: dataall.db.models.DatasetTable
+) -> dataall.db.models.ShareObjectItem:
+    yield share_item(
+        share=share2,
+        table=table2,
+        status=dataall.api.constants.ShareItemStatus.Share_Succeeded.value
+    )
+
+
+def test_init(tables1, tables2):
+    assert True
+
+
+def test_create_share_object_unauthorized(client, dataset1, env2, group2, env2group, group3):
+    """
+    Test a user that opens a share request for a group that he does not belong to. It is prevented in the frontend (client side)
+    but here we are checking that it is also unauthorized in the server side.
+    """
+
     q = """
-    mutation CreateShareObject(
-        $datasetUri:String!,
-        $input:NewShareObjectInput
-    ){
-        createShareObject(datasetUri:$datasetUri, input:$input){
-            shareUri
-            status
-            owner
-            dataset{
-                datasetUri
-                datasetName
-                exists
-            }
+      mutation CreateShareObject(
+        $datasetUri: String!
+        $itemType: String
+        $itemUri: String
+        $input: NewShareObjectInput
+      ) {
+        createShareObject(
+          datasetUri: $datasetUri
+          itemType: $itemType
+          itemUri: $itemUri
+          input: $input
+        ) {
+          shareUri
+          created
         }
-    }
+      }
     """
 
     response = client.query(
         q,
-        username=user.userName,
-        groups=[env1group],
-        datasetUri=dataset2.datasetUri,
+        username='anonymous',
+        groups=[group3.name],
+        datasetUri=dataset1.datasetUri,
         input={
-            'environmentUri': env1.environmentUri,
-            'groupUri': env1group,
-            'principalId': env1group,
-            'principalType': dataall.api.constants.PrincipalType.Group.name,
+            'environmentUri': env2.environmentUri,
+            'groupUri': env2group.groupUri,
+            'principalId': env2group.groupUri,
+            'principalType': dataall.api.constants.PrincipalType.Group.value,
         },
     )
-    return response.data.createShareObject
-
 
-def test_init(tables1, tables2):
-    assert True
+    assert 'Unauthorized' in response.errors[0].message
 
 
-def test_request_access_unauthorized(client, dataset1, env2, group2, group3):
+@pytest.mark.dependency()
```

Comment:
Regarding these test for the share api we refactored our forked version again into a setup where each test is independent of one another and instead the required state is constructed in each test with a cleanup after each tests which ensures that each test can run independent from another.

For this we first moved all the graphql queries/mutations into separate functions, so the calls can be reused in multiple test cases:

`def add_share_item(shareUri, itemUri, itemType, client, userName, groupUri):
    query = """
    mutation AddSharedItem($shareUri:String!,$input:AddSharedItemInput){
        addSharedItem(shareUri:$shareUri,input:$input){
            shareUri
            shareItemUri
            itemUri
            status
            action
        }
    }
    """

    response = client.query(
        query,
        username=userName,
        groups=[groupUri],
        shareUri=shareUri,
        input={
            "itemUri": itemUri,
            "itemType": itemType,
        },
    )

    print("Response from addSharedItem: ", response)

    return response`


We added a cleanup part to the share fixture:
`@pytest.fixture(scope="function")
def share1(env2, user, group, user2, group2, dataset1, share, client, db):
    share1 = share(
        environmentUri=env2.environmentUri, 
        userName=user2.userName, 
        groupUri=group2.groupUri, 
        datasetUri=dataset1.datasetUri
    )

    yield share1

    # Cleanup share
    # First try to direct delete
    delete_share_object_response = delete_share_object(shareUri=share1.shareUri, client=client, userName=user2.userName, groupUri=group2.groupUri)

    if delete_share_object_response.data.deleteShareObject == True:
        return

    # Then delete via revoke if still items present
    revoke_all(share1.shareUri, client, user2.userName, group2.groupUri)

    succesfull_processing_of_share_object(db, share1.shareUri)

    delete_share_object(
        shareUri=share1.shareUri,
        client=client, 
        userName=user2.userName, 
        groupUri=group2.groupUri
    )
`